java/vector/src/main/java/org/apache/arrow/vector/VectorUnloader.java

```diff
@@ -69,9 +85,79 @@ private void appendNodes(FieldVector vector, List<ArrowFieldNode> nodes, List<Ar
           vector.getField(), vector.getClass().getSimpleName(), fieldBuffers));
     }
     buffers.addAll(fieldBuffers);
+
     for (FieldVector child : vector.getChildrenFromFields()) {
       appendNodes(child, nodes, buffers);
     }
   }
 
+  // translate dictionary fields from in-memory format to message format
+  // add dictionary ids, change field types to dictionary type instead of index type
+  private void updateSchemaAndUnloadDictionaries() {
+    Map<Dictionary, ArrowDictionaryBatch> dictionaries = new HashMap<>();
+    Map<Dictionary, Long> dictionaryIds = new HashMap<>();
+
+    // go through once and collect any existing dictionary ids so that we don't duplicate them
+    for (FieldVector vector: vectors) {
+      if (vector instanceof DictionaryVector) {
+        Dictionary dictionary = ((DictionaryVector) vector).getDictionary();
+        dictionaryIds.put(dictionary, dictionary.getId());
+      }
+    }
+
+    // now generate ids for any dictionaries that haven't defined them
+    long nextDictionaryId = 0;
+    for (Entry<Dictionary, Long> entry: dictionaryIds.entrySet()) {
+      if (entry.getValue() == null) {
+        while (dictionaryIds.values().contains(nextDictionaryId)) {
+          nextDictionaryId++;
+        }
+        dictionaryIds.put(entry.getKey(), nextDictionaryId);
+      }
+    }
+
+    // go through again to add dictionary id to the schema fields and to unload the dictionary batches
+    for (FieldVector vector: vectors) {
+      if (vector instanceof DictionaryVector) {
+        Dictionary dictionary = ((DictionaryVector) vector).getDictionary();
+        long dictionaryId = dictionaryIds.get(dictionary);
+        Field field = vector.getField();
+        // find the dictionary field in the schema
```

Comment:
as @wesm mentioned, we didn't want to assign dictionary ids until encoding for transfer. Having them defined up front should make this method simpler